RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc

```diff
@@ -0,0 +1,123 @@
+#include "RecoTracker/FinalTrackSelectors/interface/TrackMVAClassifier.h"
+
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+
+#include "getBestVertex.h"
+
+#include "TrackingTools/Records/interface/TfGraphRecord.h"
+#include "PhysicsTools/TensorFlow/interface/TensorFlow.h"
+#include "RecoTracker/FinalTrackSelectors/interface/TfGraphDefWrapper.h"
+
+struct TfDnnCache {
+  TfDnnCache() : session_(nullptr) {}
+  tensorflow::Session* session_;
+};
+
+namespace {
+  struct tfDnn {
+    tfDnn(const edm::ParameterSet& cfg)
+        : tfDnnLabel_(cfg.getParameter<std::string>("tfDnnLabel"))
+
+    {}
+    TfDnnCache* cache_ = new TfDnnCache();
+
+    ~tfDnn() {
+      if (cache_->session_) {
+        tensorflow::closeSession(cache_->session_);
+        delete cache_;
+      }
+    }
+
+    static const char* name() { return "TrackTfClassifier"; }
+
+    static void fillDescriptions(edm::ParameterSetDescription& desc) {
+      desc.add<std::string>("tfDnnLabel", "trackSelectionTf");
+    }
+
+    void beginStream() {}
+
+    void initEvent(const edm::EventSetup& es) {
+      if (!cache_->session_) {
+        edm::ESHandle<TfGraphDefWrapper> tfDnnHandle;
+        es.get<TfGraphRecord>().get(tfDnnLabel_, tfDnnHandle);
+        tensorflow::GraphDef* graphDef_ = tfDnnHandle.product()->GetGraphDef();
+        cache_->session_ = tensorflow::createSession(
```

Comment:
ah, ok, this creates a new session per stream.
This explains the scaling.

Reply:
Ah, yes, I could have pointed you to that, sorry. I thought you were asking why it was implemented that way.

Reply:
we need `cache_` to be a part of the global cache, so the base `class TrackMVAClassifierBase : public edm::stream::EDProducer<> {` needs some modification

Reply:
I thought Session is not thread safe, and therefore may not be shared across stream instances.

Reply:
all other TF sessions that we have are in streams; although I'm still not sure if this was done deliberately due to the session being not MT safe.
@Dr15Jones @riga is a session not thread safe?